dev: update eth_getBlockTransactionCountByHash & byNumber to return KKT txs count

dpinones opened the pull request:
<!--- Please provide a general summary of your changes in the title above -->
The `eth_getBlockTransactionCountByHash` and `eth_getBlockTransactionCountByNumber` functions get the number of transactions per block from the `starknet_getBlockTransactionCount` function. Previously they used the `starknet_getBlockWithTxHashes` function.

Updated issue resolution: https://github.com/sayajin-labs/kakarot-rpc/pull/108#issuecomment-1444718187

<!-- Give an estimate of the time you spent on this PR in terms of work days. Did you spend 0.5 days on this PR or rather 2 days?  -->

Time spent on this PR: 1.5 day

<!-- Which issue is this PR related to? -->

Closes #92 

# Pull Request type

<!-- Please try to limit your pull request to one type; submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A

# What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

-

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this does introduce a breaking change, please describe the impact and migration path for existing applications below. -->

# Other information

<!-- Any other information that is important to this PR, such as screenshots of how the component looks before and after the change. -->

dpinones commented:
It only remains to update the tests

Eikix reviewed (changes requested):
thanks for your PR, 
I realize the issue was not perfectly clear: currently, this PR counts the number of transactions on Starknet at a certain block,
that being said, our business logic is to return the number of Kakarot transactions, which is a subset of the number of starknet transactions at a certain block.



This PR will need to be changed altered in order to return this specific number.


Changes: 
A - Change the logic so that it counts the number of Kakarot (KKT) tx, you can draw inspiration from the `starknet_tx_into_kakarot_tx` function and `filter_starknet_into_eth_txs`, since a Starknet transaction is considered (for now) a KKT tx if and only if the class hash of the sender address corresponds to the account_proxy class hash (constant or env var) from Kakarot.

B - Could you rebase it on the new state of main 

C -  add a test for your use case?

you can draw inspiration in `rpc/tests/rpc`

1. add a wiremock route for your Starknet call get_transaction_count in `core/src/utils/wiremock_utils`
2. add a test in `rpc/tests/rpc`

dpinones commented:
If I understood correctly, to meet the requirements of this PR, we cannot use the `get_transaction_count` function since it only returns a number representing the transaction count, and we need detailed information on each transaction to filter out only the KKT transactions.

Therefore, I made the following changes:

1. I used the `get_block_with_txs` function to obtain all transactions in a block.
2. I used the `filter_starknet_into_eth_txs` function to filter the transactions and obtain only the KKT transactions.
3. Finally, I return the count of filtered transactions that are KKT transactions.
4. I also added test cases and performed a rebase with the current state of the main branch.
5. Additionally, I added a mock for the `get_transaction_count` function, although I did not use it.

Please let me know if there are any corrections that need to be made.

Eikix commented on core/src/utils/wiremock_utils.rs:
```diff
@@ -126,6 +135,46 @@ pub async fn setup_wiremock() -> String {
+        .respond_with(
+            ResponseTemplate::new(200)
+                .set_body_raw(
+                    include_str!("data/blocks/starknet_getBlockTransactionCount.json"),
```
did you add this file?

since it's not a `starknet_getBlockTransactionCount` call we're doing inside our core logic, we don't have to call it that, we don't even have to mock it,

what we're really mocking is `client.get_block_with_txs` from [this line](https://github.com/sayajin-labs/kakarot-rpc/pull/108/files#diff-e6d20c367d720630c43945964dc7fdb3105ee5c321f359ce26f51c44b820f68aR449)

which means we have to "format" well a call to get_block_with_txs if it's not already being mocked properly,
if tests pass etc. maybe an already existing get_block_with_txs mock is enough for our use case of filtering txs

dpinones replied:
You were right, the mock was eliminated because it was not being used.

Eikix reviewed (changes requested):
very cool PR, thanks for being reactive and understanding,
almost ready to merge 🔥

Eikix commented on core/src/utils/wiremock_utils.rs:
```diff
@@ -44,6 +44,15 @@ impl<'a, StarknetParams> StarknetRpcBaseData<'a, StarknetParams> {
         }
     }
 
+    pub fn get_block_transaction_count(params: StarknetParams) -> Self {
```
do you still need this?

dpinones replied:
You were right, I don't need it anymore

Eikix commented on docs/methods/eth_getBlockTransactionCountByNumber.md:
```diff
@@ -30,7 +30,7 @@ transactions in a block matching the given block number.
 ### Starknet methods
 
-- [starknet_getBlockWithTxs](https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_api_openrpc.json#L44)
+- [starknet_getBlockTransactionCount](https://github.com/starkware-libs/starknet-specs/blob/a789ccc3432c57777beceaa53a34a7ae2f25fda0/api/starknet_api_openrpc.json#L373)
```
I think we're not using it in the end

Eikix commented on rpc/tests/rpc.rs:
```diff
@@ -224,15 +224,14 @@ mod tests {
         let kakarot_rpc = setup_kakarot_eth_rpc().await;
         let hash =
-            H256::from_str("0x0197be2810df6b5eedd5d9e468b200d0b845b642b81a44755e19047f08cc8c6e")
+            H256::from_str("0x0449aa33ad836b65b10fa60082de99e24ac876ee2fd93e723a99190a530af0a9")
```
why are we changing the hash btw? what does it bring?

dpinones replied:
Because now I am using the "get_block_with_txs" method and in the wiremock_utils file it has that [hash](https://github.com/sayajin-labs/kakarot-rpc/blob/abfb3d5564b930782a153fe4bfdb770dc59cd8a6/core/src/utils/wiremock_utils.rs#L131)

Eikix reviewed (approved):
lgtm 🔥🔥🔥🔥🔥🔥🔥
